OpenRA.Mods.Common/Traits/Attack/AttackBase.cs

```diff
@@ -211,12 +208,21 @@ public bool HasAnyValidWeapons(Target t)
 
 			// PERF: Avoid LINQ.
 			foreach (var armament in Armaments)
-				if (!armament.OutOfAmmo && armament.Weapon.IsValidAgainst(t, self.World, self))
+			{
+				// If 'onlyTargetCenterWeapons' is true, we don't care about current ammo state, only Weapon.TargetActorCenter
+				var checkIsValidAgainst = onlyTargetCenterWeapons ? armament.Weapon.TargetActorCenter : !armament.OutOfAmmo;
+				if (checkIsValidAgainst && armament.Weapon.IsValidAgainst(t, self.World, self))
 					return true;
+			}
 
 			return false;
 		}
 
+		public virtual WPos GetTargetPosition(WPos pos, Target target, bool onlyTargetCenterWeapons = false)
+		{
+			return HasAnyValidWeapons(target, onlyTargetCenterWeapons) ? target.CenterPosition : target.Positions.PositionClosestTo(pos);
```

Comment:
I agree that using `HasAnyValidWeapons` with that boolean makes the code harder to understand, but it works as intended in game, as far as I could tell while testing.

But I don't really like that boolean arg, in my opinion the code will be cleaner and easier to read if I separate `HasAnyValidWeapons(..., true)` into its own `HasAnyValidCenterTargetingWeapons(Target t)` method, to get rid of `onlyTargetCenterWeapons` and the TargetActorCenter/ammo gymnastics.